File: src/server/core/typescript/src/page/build-loader.ts
/* src/server/core/typescript/src/page/build-loader.ts */

import { existsSync, readFileSync } from 'node:fs'
import { join } from 'node:path'
import type { PageDef, PageAssets, LayoutDef, LoaderFn, LoaderResult, I18nConfig } from './index.js'
import type { RpcHashMap } from '../http.js'

interface RouteManifest {
	layouts?: Record<string, LayoutManifestEntry>
	routes: Record<string, RouteManifestEntry>
	data_id?: string
	i18n?: {
		locales: string[]
		default: string
		mode?: string
		cache?: boolean
		route_hashes?: Record<string, string>
		content_hashes?: Record<string, Record<string, string>>
	}
}

interface LayoutManifestEntry {
	template?: string
	templates?: Record<string, string>
	loaders?: Record<string, LoaderConfig>
	parent?: string
	i18n_keys?: string[]
}

interface RouteManifestEntry {
	template?: string
	templates?: Record<string, string>
	layout?: string
	loaders: Record<string, LoaderConfig>
	head_meta?: string
	i18n_keys?: string[]
	assets?: PageAssets
	projections?: Record<string, string[]>
	prerender?: boolean
}

interface LoaderConfig {
	procedure: string
	params?: Record<string, string | ParamConfig>
	handoff?: 'client'
}

interface ParamConfig {
	from: 'route' | 'query'
	type?: 'string' | 'int'
}

function normalizeParamConfig(value: string | ParamConfig): ParamConfig {
	return typeof value === 'string' ? { from: value as ParamConfig['from'] } : value
}

function buildLoaderFn(config: LoaderConfig): LoaderFn {
	return (params, searchParams): LoaderResult => {
		const input: Record<string, unknown> = {}
		if (config.params) {
			for (const [key, raw_mapping] of Object.entries(config.params)) {
				const mapping = normalizeParamConfig(raw_mapping)
				const raw = mapping.from === 'query' ? (searchParams?.get(key) ?? undefined) : params[key]
				if (raw !== undefined) {
					input[key] = mapping.type === 'int' ? Number(raw) : raw
				}
			}
		}
		return { procedure: config.procedure, input }
	}
}

function buildLoaderFns(configs: Record<string, LoaderConfig>): Record<string, LoaderFn> {
	const fns: Record<string, LoaderFn> = {}
	for (const [key, config] of Object.entries(configs)) {
		fns[key] = buildLoaderFn(config)
	}
	return fns
}

function resolveTemplatePath(
	entry: { template?: string; templates?: Record<string, string> },
	defaultLocale: string | undefined,
): string {
	if (entry.template) return entry.template
	if (entry.templates) {
		const locale = defaultLocale ?? (Object.keys(entry.templates)[0] as string)
		const path = entry.templates[locale]
		if (!path) throw new Error(`No template for locale "${locale}"`)
		return path
	}
	throw new Error("Manifest entry has neither 'template' nor 'templates'")
}

/** Load all locale templates for a manifest entry, keyed by locale */
function loadLocaleTemplates(
	entry: { templates?: Record<string, string> },
	distDir: string,
): Record<string, string> | undefined {
	if (!entry.templates) return undefined
	const result: Record<string, string> = {}
	for (const [locale, relPath] of Object.entries(entry.templates)) {
		result[locale] = readFileSync(join(distDir, relPath), 'utf-8')
	}
	return result
}

/** Callback that returns template + localeTemplates for a layout entry */
type TemplateGetter = (
	id: string,
	entry: LayoutManifestEntry,
) => { template: string; localeTemplates?: Record<string, string> }

/** Resolve parent chain for a layout, returning outer-to-inner order */
function resolveLayoutChain(
	layoutId: string,
	layoutEntries: Record<string, LayoutManifestEntry>,
	getTemplates: TemplateGetter,
): LayoutDef[] {
	const chain: LayoutDef[] = []
	let currentId: string | undefined = layoutId

	while (currentId) {
		const entry: LayoutManifestEntry | undefined = layoutEntries[currentId]
		if (!entry) break
		const { template, localeTemplates } = getTemplates(currentId, entry)
		chain.push({
			id: currentId,
			template,
			localeTemplates,
			loaders: buildLoaderFns(entry.loaders ?? {}),
		})
		currentId = entry.parent
	}

	// Reverse: we walked inner->outer, but want outer->inner
	chain.reverse()
	return chain
}

/** Create a proxy object that lazily reads locale templates from disk */
function makeLocaleTemplateGetters(
	templates: Record<string, string>,
	distDir: string,
): Record<string, string> {
	const obj: Record<string, string> = {}
	for (const [locale, relPath] of Object.entries(templates)) {
		const fullPath = join(distDir, relPath)
		Object.defineProperty(obj, locale, {
			get: () => readFileSync(fullPath, 'utf-8'),
			enumerable: true,
		})
	}
	return obj
}

/** Merge i18n_keys from route + layout chain into a single list */
function mergeI18nKeys(
	route: RouteManifestEntry,
	layoutEntries: Record<string, LayoutManifestEntry>,
): string[] | undefined {
	const keys: string[] = []
	if (route.layout) {
		let currentId: string | undefined = route.layout
		while (currentId) {
			const entry: LayoutManifestEntry | undefined = layoutEntries[currentId]
			if (!entry) break
			if (entry.i18n_keys) keys.push(...entry.i18n_keys)
			currentId = entry.parent
		}
	}
	if (route.i18n_keys) keys.push(...route.i18n_keys)
	return keys.length > 0 ? keys : undefined
}

export interface BuildOutput {
	pages: Record<string, PageDef>
	rpcHashMap: RpcHashMap | undefined
	i18n: I18nConfig | null
}

/** Load all build artifacts (pages, rpcHashMap, i18n) in one call */
export function loadBuild(distDir: string): BuildOutput {
	return {
		pages: loadBuildOutput(distDir),
		rpcHashMap: loadRpcHashMap(distDir),
		i18n: loadI18nMessages(distDir),
	}
}

/** Load all build artifacts with lazy template getters (for dev mode) */
export function loadBuildDev(distDir: string): BuildOutput {
	return {
		pages: loadBuildOutputDev(distDir),
		rpcHashMap: loadRpcHashMap(distDir),
		i18n: loadI18nMessages(distDir),
	}
}

/** Load the RPC hash map from build output (returns undefined when obfuscation is off) */
export function loadRpcHashMap(distDir: string): RpcHashMap | undefined {
	const hashMapPath = join(distDir, 'rpc-hash-map.json')
	try {
		return JSON.parse(readFileSync(hashMapPath, 'utf-8')) as RpcHashMap
	} catch {
		return undefined
	}
}

/** Load i18n config and messages from build output */
export function loadI18nMessages(distDir: string): I18nConfig | null {
	const manifestPath = join(distDir, 'route-manifest.json')
	try {
		const manifest = JSON.parse(readFileSync(manifestPath, 'utf-8')) as RouteManifest
		if (!manifest.i18n) return null

		const mode = (manifest.i18n.mode ?? 'memory') as 'memory' | 'paged'
		const cache = manifest.i18n.cache ?? false
		const routeHashes = manifest.i18n.route_hashes ?? {}
		const contentHashes = manifest.i18n.content_hashes ?? {}

		// Memory mode: preload all route messages per locale
		// Paged mode: store distDir for on-demand reads
		const messages: Record<string, Record<string, Record<string, string>>> = {}
		if (mode === 'memory') {
			const i18nDir = join(distDir, 'i18n')
			for (const locale of manifest.i18n.locales) {
				const localePath = join(i18nDir, `${locale}.json`)
				if (existsSync(localePath)) {
					messages[locale] = JSON.parse(readFileSync(localePath, 'utf-8')) as Record<
						string,
						Record<string, string>
					>
				} else {
					messages[locale] = {}
				}
			}
		}

		return {
			locales: manifest.i18n.locales,
			default: manifest.i18n.default,
			mode,
			cache,
			routeHashes,
			contentHashes,
			messages,
			distDir: mode === 'paged' ? distDir : undefined,
		}
	} catch {
		return null
	}
}

export function loadBuildOutput(distDir: string): Record<string, PageDef> {
	const manifestPath = join(distDir, 'route-manifest.json')
	const raw = readFileSync(manifestPath, 'utf-8')
	const manifest = JSON.parse(raw) as RouteManifest
	const defaultLocale = manifest.i18n?.default

	// Load layout templates (default + all locales)
	const layoutTemplates: Record<string, string> = {}
	const layoutLocaleTemplates: Record<string, Record<string, string>> = {}
	const layoutEntries = manifest.layouts ?? {}
	for (const [id, entry] of Object.entries(layoutEntries)) {
		layoutTemplates[id] = readFileSync(
			join(distDir, resolveTemplatePath(entry, defaultLocale)),
			'utf-8',
		)
		const lt = loadLocaleTemplates(entry, distDir)
		if (lt) layoutLocaleTemplates[id] = lt
	}

	// Static dir for prerendered pages (adjacent to output dir)
	const staticDir = join(distDir, '..', 'static')
	const hasStaticDir = existsSync(staticDir)

	const pages: Record<string, PageDef> = {}
	for (const [path, entry] of Object.entries(manifest.routes)) {
		const templatePath = join(distDir, resolveTemplatePath(entry, defaultLocale))
		const template = readFileSync(templatePath, 'utf-8')

		const loaders = buildLoaderFns(entry.loaders)
		const layoutChain = entry.layout
			? resolveLayoutChain(entry.layout, layoutEntries, (id) => ({
					template: layoutTemplates[id] ?? '',
					localeTemplates: layoutLocaleTemplates[id],
				}))
			: []

		// Merge i18n_keys from layout chain + route
		const i18nKeys = mergeI18nKeys(entry, layoutEntries)

		const page: PageDef = {
			template,
			localeTemplates: loadLocaleTemplates(entry, distDir),
			loaders,
			layoutChain,
			headMeta: entry.head_meta,
			dataId: manifest.data_id,
			i18nKeys,
			pageAssets: entry.assets,
			projections: entry.projections,
		}

		if (entry.prerender && hasStaticDir) {
			page.prerender = true
			page.staticDir = staticDir
		}

		pages[path] = page
	}
	return pages
}

/** Load build output with lazy template getters -- templates re-read from disk on each access */
export function loadBuildOutputDev(distDir: string): Record<string, PageDef> {
	const manifestPath = join(distDir, 'route-manifest.json')
	const raw = readFileSync(manifestPath, 'utf-8')
	const manifest = JSON.parse(raw) as RouteManifest
	const defaultLocale = manifest.i18n?.default

	const layoutEntries = manifest.layouts ?? {}

	const pages: Record<string, PageDef> = {}
	for (const [path, entry] of Object.entries(manifest.routes)) {
		const templatePath = join(distDir, resolveTemplatePath(entry, defaultLocale))
		const loaders = buildLoaderFns(entry.loaders)
		const layoutChain = entry.layout
			? resolveLayoutChain(entry.layout, layoutEntries, (id, layoutEntry) => {
					const tmplPath = join(distDir, resolveTemplatePath(layoutEntry, defaultLocale))
					const def = {
						template: '',
						localeTemplates: layoutEntry.templates
							? makeLocaleTemplateGetters(layoutEntry.templates, distDir)
							: undefined,
					}
					Object.defineProperty(def, 'template', {
						get: () => readFileSync(tmplPath, 'utf-8'),
						enumerable: true,
					})
					return def
				})
			: []

		const localeTemplates = entry.templates
			? makeLocaleTemplateGetters(entry.templates, distDir)
			: undefined

		// Merge i18n_keys from layout chain + route
		const i18nKeys = mergeI18nKeys(entry, layoutEntries)

		const page: PageDef = {
			template: '', // placeholder, overridden by getter
			localeTemplates,
			loaders,
			layoutChain,
			dataId: manifest.data_id,
			i18nKeys,
			pageAssets: entry.assets,
			projections: entry.projections,
		}
		Object.defineProperty(page, 'template', {
			get: () => readFileSync(templatePath, 'utf-8'),
			enumerable: true,
		})
		pages[path] = page
	}
	return pages
}

File: src/server/core/go/build_loader.go
/* src/server/core/go/build_loader.go */

// Load page definitions from seam build output on disk.
// Reads route-manifest.json, loads templates, constructs PageDef with loaders.

package seam

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

type routeManifest struct {
	Layouts map[string]layoutEntry `json:"layouts"`
	Routes  map[string]routeEntry  `json:"routes"`
	DataID  string                 `json:"data_id"`
	I18n    *i18nManifest          `json:"i18n"`
}

type i18nManifest struct {
	Locales       []string                     `json:"locales"`
	Default       string                       `json:"default"`
	Mode          string                       `json:"mode"`
	Cache         bool                         `json:"cache"`
	RouteHashes   map[string]string            `json:"route_hashes"`
	ContentHashes map[string]map[string]string `json:"content_hashes"`
}

type layoutEntry struct {
	Template  string            `json:"template"`
	Templates map[string]string `json:"templates"`
	Loaders   json.RawMessage   `json:"loaders"`
	Parent    string            `json:"parent"`
	I18nKeys  []string          `json:"i18n_keys"`
}

type routeEntry struct {
	Template    string              `json:"template"`
	Templates   map[string]string   `json:"templates"`
	Layout      string              `json:"layout"`
	Loaders     json.RawMessage     `json:"loaders"`
	HeadMeta    string              `json:"head_meta"`
	I18nKeys    []string            `json:"i18n_keys"`
	Assets      *PageAssets         `json:"assets"`
	Projections map[string][]string `json:"projections"`
}

// pickTemplate returns the template path: prefer singular "template",
// fall back to default locale, then any first value from "templates".
func pickTemplate(single string, multi map[string]string, defaultLocale string) string {
	if single != "" {
		return single
	}
	if multi != nil {
		if defaultLocale != "" {
			if t, ok := multi[defaultLocale]; ok {
				return t
			}
		}
		for _, t := range multi {
			return t
		}
	}
	return ""
}

type loaderConfig struct {
	Procedure string                     `json:"procedure"`
	Params    map[string]loaderParamConf `json:"params"`
}

type loaderParamConf struct {
	From string `json:"from"`
	Type string `json:"type"`
}

func parseLoaders(raw json.RawMessage) []LoaderDef {
	if len(raw) == 0 || string(raw) == "null" {
		return nil
	}
	var obj map[string]loaderConfig
	if err := json.Unmarshal(raw, &obj); err != nil {
		return nil
	}

	var loaders []LoaderDef
	for dataKey, cfg := range obj {
		proc := cfg.Procedure
		params := cfg.Params
		loaders = append(loaders, LoaderDef{
			DataKey:   dataKey,
			Procedure: proc,
			InputFn:   buildInputFn(params),
		})
	}
	return loaders
}

func buildInputFn(params map[string]loaderParamConf) func(map[string]string) any {
	return func(routeParams map[string]string) any {
		obj := make(map[string]any)
		for key, cfg := range params {
			if cfg.From == "route" {
				obj[key] = routeParams[key]
			}
		}
		return obj
	}
}

// resolveLayoutChain walks from child to root, nesting page content inside layout templates.
func resolveLayoutChain(layoutID, pageTemplate string, layouts map[string]layoutResolved) string {
	result := pageTemplate
	current := layoutID

	for current != "" {
		lr, ok := layouts[current]
		if !ok {
			break
		}
		result = strings.Replace(lr.template, "<!--seam:outlet-->", result, 1)
		current = lr.parent
	}

	return result
}

type layoutResolved struct {
	template string
	parent   string
}

// RpcHashMap maps hashed procedure names back to originals.
type RpcHashMap struct {
	Salt       string            `json:"salt"`
	Batch      string            `json:"batch"`
	Procedures map[string]string `json:"procedures"`
}

// ReverseLookup builds hash -> original name map.
func (m *RpcHashMap) ReverseLookup() map[string]string {
	rev := make(map[string]string, len(m.Procedures))
	for name, hash := range m.Procedures {
		rev[hash] = name
	}
	return rev
}

// LoadRpcHashMap loads the RPC hash map from build output (returns nil when not present).
func LoadRpcHashMap(dir string) *RpcHashMap {
	data, err := os.ReadFile(filepath.Join(dir, "rpc-hash-map.json"))
	if err != nil {
		return nil
	}
	var m RpcHashMap
	if err := json.Unmarshal(data, &m); err != nil {
		return nil
	}
	return &m
}

// LoadBuildOutput loads page definitions from seam build output on disk.
func LoadBuildOutput(dir string) ([]PageDef, error) {
	manifestPath := filepath.Join(dir, "route-manifest.json")
	data, err := os.ReadFile(manifestPath)
	if err != nil {
		return nil, fmt.Errorf("read route-manifest.json: %w", err)
	}

	var manifest routeManifest
	if err := json.Unmarshal(data, &manifest); err != nil {
		return nil, fmt.Errorf("parse route-manifest.json: %w", err)
	}

	defaultLocale := ""
	if manifest.I18n != nil {
		defaultLocale = manifest.I18n.Default
	}

	// Load layout templates (default locale)
	layouts := make(map[string]layoutResolved)
	for id, entry := range manifest.Layouts {
		tmplPath := pickTemplate(entry.Template, entry.Templates, defaultLocale)
		if tmplPath == "" {
			continue
		}
		tmplBytes, err := os.ReadFile(filepath.Join(dir, tmplPath))
		if err != nil {
			return nil, fmt.Errorf("read layout template %s: %w", tmplPath, err)
		}
		layouts[id] = layoutResolved{template: string(tmplBytes), parent: entry.Parent}
	}

	// Load layout templates per locale for locale-specific resolution
	// layoutLocaleTemplates[locale][layoutID] = layoutResolved
	layoutLocaleTemplates := make(map[string]map[string]layoutResolved)
	if manifest.I18n != nil {
		for id, entry := range manifest.Layouts {
			if entry.Templates == nil {
				continue
			}
			for locale, tmplPath := range entry.Templates {
				tmplBytes, err := os.ReadFile(filepath.Join(dir, tmplPath))
				if err != nil {
					return nil, fmt.Errorf("read layout locale template %s: %w", tmplPath, err)
				}
				if layoutLocaleTemplates[locale] == nil {
					layoutLocaleTemplates[locale] = make(map[string]layoutResolved)
				}
				layoutLocaleTemplates[locale][id] = layoutResolved{
					template: string(tmplBytes),
					parent:   entry.Parent,
				}
			}
		}
	}

	var pages []PageDef

	for routePath, entry := range manifest.Routes {
		tmplPath := pickTemplate(entry.Template, entry.Templates, defaultLocale)
		if tmplPath == "" {
			continue
		}
		tmplBytes, err := os.ReadFile(filepath.Join(dir, tmplPath))
		if err != nil {
			return nil, fmt.Errorf("read route template %s: %w", tmplPath, err)
		}
		pageTemplate := string(tmplBytes)

		// Resolve layout chain
		template := pageTemplate
		if entry.Layout != "" {
			template = resolveLayoutChain(entry.Layout, pageTemplate, layouts)
		}

		// Build locale-specific pre-resolved templates when i18n is active
		var localeTemplates map[string]string
		if manifest.I18n != nil && entry.Templates != nil {
			localeTemplates = make(map[string]string)
			for locale, ltPath := range entry.Templates {
				ltBytes, err := os.ReadFile(filepath.Join(dir, ltPath))
				if err != nil {
					return nil, fmt.Errorf("read route locale template %s: %w", ltPath, err)
				}
				pageTmpl := string(ltBytes)
				resolved := pageTmpl
				if entry.Layout != "" {
					localeLayouts := layoutLocaleTemplates[locale]
					if localeLayouts == nil {
						localeLayouts = layouts
					}
					resolved = resolveLayoutChain(entry.Layout, pageTmpl, localeLayouts)
				}
				localeTemplates[locale] = resolved
			}
		}

		// Collect loaders: layout chain loaders + route loaders
		// Also build layout chain with per-layout loader key assignments
		var allLoaders []LoaderDef
		var layoutChain []LayoutChainEntry
		if entry.Layout != "" {
			current := entry.Layout
			for current != "" {
				if le, ok := manifest.Layouts[current]; ok {
					layoutLoaders := parseLoaders(le.Loaders)
					var loaderKeys []string
					for _, ld := range layoutLoaders {
						loaderKeys = append(loaderKeys, ld.DataKey)
					}
					layoutChain = append(layoutChain, LayoutChainEntry{ID: current, LoaderKeys: loaderKeys})
					allLoaders = append(allLoaders, layoutLoaders...)
					current = le.Parent
				} else {
					break
				}
			}
			// Reverse: walked inner->outer, want outer->inner (matching TS)
			for i, j := 0, len(layoutChain)-1; i < j; i, j = i+1, j-1 {
				layoutChain[i], layoutChain[j] = layoutChain[j], layoutChain[i]
			}
		}
		pageLoaders := parseLoaders(entry.Loaders)
		var pageLoaderKeys []string
		for _, ld := range pageLoaders {
			pageLoaderKeys = append(pageLoaderKeys, ld.DataKey)
		}
		allLoaders = append(allLoaders, pageLoaders...)

		dataID := manifest.DataID
		if dataID == "" {
			dataID = "__data"
		}
		// Merge i18n_keys from layout chain + route
		var i18nKeys []string
		if entry.Layout != "" {
			current := entry.Layout
			for current != "" {
				if le, ok := manifest.Layouts[current]; ok {
					i18nKeys = append(i18nKeys, le.I18nKeys...)
					current = le.Parent
				} else {
					break
				}
			}
		}
		i18nKeys = append(i18nKeys, entry.I18nKeys...)

		pages = append(pages, PageDef{
			Route:           routePath,
			Template:        template,
			LocaleTemplates: localeTemplates,
			Loaders:         allLoaders,
			DataID:          dataID,
			LayoutChain:     layoutChain,
			PageLoaderKeys:  pageLoaderKeys,
			I18nKeys:        i18nKeys,
			HeadMeta:        entry.HeadMeta,
			Assets:          entry.Assets,
			Projections:     entry.Projections,
		})
	}

	return pages, nil
}

// LoadI18nConfig loads i18n configuration and locale messages from build output.
// Returns nil when i18n is not configured.
func LoadI18nConfig(dir string) *I18nConfig {
	manifestData, err := os.ReadFile(filepath.Join(dir, "route-manifest.json"))
	if err != nil {
		return nil
	}
	var manifest routeManifest
	if err := json.Unmarshal(manifestData, &manifest); err != nil {
		return nil
	}
	if manifest.I18n == nil || len(manifest.I18n.Locales) == 0 {
		return nil
	}

	i18n := manifest.I18n
	mode := i18n.Mode
	if mode == "" {
		mode = "memory"
	}

	// Memory mode: preload route-keyed messages per locale from i18n/{locale}.json
	// Paged mode: store distDir for on-demand reads
	messages := make(map[string]map[string]json.RawMessage)
	distDir := ""

	if mode == "memory" {
		i18nDir := filepath.Join(dir, "i18n")
		for _, locale := range i18n.Locales {
			localePath := filepath.Join(i18nDir, locale+".json")
			data, err := os.ReadFile(localePath)
			if err != nil {
				messages[locale] = make(map[string]json.RawMessage)
				continue
			}
			var routeMessages map[string]json.RawMessage
			if err := json.Unmarshal(data, &routeMessages); err != nil {
				messages[locale] = make(map[string]json.RawMessage)
				continue
			}
			messages[locale] = routeMessages
		}
	} else {
		distDir = dir
	}

	return &I18nConfig{
		Locales:       i18n.Locales,
		Default:       i18n.Default,
		Mode:          mode,
		Cache:         i18n.Cache,
		RouteHashes:   i18n.RouteHashes,
		ContentHashes: i18n.ContentHashes,
		Messages:      messages,
		DistDir:       distDir,
	}
}
